platform/stm32f0/drivers/componolit-runtime-drivers-gpio.adb

```diff
@@ -41,27 +41,82 @@ is
       Async_Writers,
       Effective_Writes;
 
-   Shadow_Config : Configuration := Config_Registers;
+   Shadow_Config : Pull_Config := (others => (others => Port_In));
+
+   Modes : Proof_Pin_Mode := (others => Port_In) with Ghost;
+
+   Pins : Configured_Pins := (others => False) with Ghost;
+
+   --  Proof procedure to help gnatprove unrolling the loop
+   --  over all pins.
+   procedure Prf_Unfold_Valid with
+      Ghost,
+      Pre  => (for all Pn in Pin => Valid (Pn)),
+      Post => (for all Pn in Pin =>
+                  Modes (Pn) =
+                     Shadow_Config (Bank_Select (Pn)) (Pin_Offset (Pn)));
+
+   procedure Prf_Unfold_Valid
+   is
+   begin
+      for P in Pin loop
+         pragma Assert (Valid (P));
+         pragma Loop_Invariant (for all Q in Pin'First .. P =>
+                                   Modes (Q) =
+                                      Shadow_Config (Bank_Select (Q))
+                                         (Pin_Offset (Q)));
```

Comment:
You should put `(Bank_Select (Q))` onto a separate line (like `(Pin_Offset (Q))`). With this layout I still find it challenging to grasp that both are indices into a two-dimensional array.

Reply:
Fixed by https://github.com/Componolit/ada-runtime/pull/73/commits/946677eb98d1ce3f6baca472fc40a48c990bc090.